exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,824 @@
+
+### Allowing different kinds of functions
+
+The syntax could also distinguish multiple kinds of functions,
+some of which are composable and some not.
+
+Or, the function registry could allow functions to be declared in this way,
+with incorrect uses of functions being resolution errors.
+
+(See "Different kinds of composition" under "Alternatives to consider".)
+
+### Summarizing use cases
+
+There seem to be two areas of ambiguity:
+
+* Are named values essentially strings with metadata,
+or are they structured? (Model 1 vs. Model 2)
+* In Model 2, some functions "look back" for the original value,
+(like `number`)
+while others return a new "source value"
+(like `getAge` in Example B1).
+
+The choice of internal value influences both areas,
+or rather, the desired answers to these questions
+constrain the choice of internal value.
+
+Composition might mean that "the second function operates on the output of the first",
+or "the second function operates on the **input** of the first, plus 'hints' supplied by the first",
+or it might mean either one depending on which function(s) are involved.
+
+The question is how to craft the spec in a way that is consistent with expectations.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+In the rest of this document, we assume some version of Model 2.
+However, if Model 1 is more desired, the questions arise of how to
+forbid compositions of functions that would do surprising things
+under that model.
+
+Even under Model 2, some instances of composition won't make sense,
+so we need to define the error behavior when it doesn't make sense.
+
+This implies that we need:
+
+* Preservation of options
+* Preservation of the original operand value, at least for some functions
+* Passing through a new operand value _and_ options when appropriate, as with functions that extract fields
+(see Example B1)
+* (Possibly) Preservation of data about the names of functions that were previously called.
+
+The function registry needs to be explicit about which pieces of data
+a function preserves,
+and what it expects in its input.
+
+More generally than just "preservation of options",
+a solution needs to specify a minimum set of requirements
+for the internal value representation,
+so that functions can be passed the values they need.
+It also needs to provide a mechanism for declaring
+when functions can compose with each other.
+
+Other requirements:
+
+### Avoid over-constraining implementations
+
+An implementation is free to use any types
+and set of operations for coercing between types
+as long as it preserves the observable behavior
+of a message.
+
+The proposed solution should not impose
+unnecessary constraints on the implementation.
+However, it _should_ make the requirements
+for "resolved values"
+clear enough so that implementors can
+make a well-informed decision
+on what these types and operations should be.
```

Comment:
Is this the expected primary outcome of this work? As in, normative text or examples of (custom) functions operating on values with some expected results?

Reply:
The answer is that I'm not sure, and part of the goal of writing a doc that doesn't present a proposed solution is to narrow down what the outcome should be: normative changes and/or a set of use cases that must be supported (which would also be normative, but would be different from the style of the existing spec.) I was hoping that would happen as part of the discussion of this design doc.

I added some text along those lines in [cfa882a](https://github.com/unicode-org/message-format-wg/pull/753/commits/cfa882a3b9eb6fed277b8c847c5ccefb7918b46c). I also added some text at the end about different directions that a proposed design could go in. If there are any broad categories of solution that I left out, I'd like to include them. 

I think that to answer your question, we have to pick one of those general directions.